presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -674,6 +774,209 @@ public String toString()
+                    handle.getTableName(),
+                    handle.getConnectorSession());
+            ConnectorPartitionResult result = splitManager.getPartitions(hiveTableHandle,
+                    TupleDomain.<ConnectorColumnHandle>all());
+            List<ConnectorPartition> partitions = result.getPartitions();
+            Iterator<ConnectorPartition> pIter = partitions.iterator();
+
+            while (pIter.hasNext()) {
+                HivePartition p = (HivePartition) pIter.next();
+                // partitionId will be of form pKey1=value1/pKey2=value2
+                String partitionId = p.getPartitionId();
+
+                if (!partitionsKnown.add(partitionId.toString())) {
+                    throw new RuntimeException(String.format("Table '%s' has duplicate partitions for '%s'", handle.getTableName(), partitionId));
+                }
+            }
+
+            Table table = getMetastore().getTable(handle.getSchemaName(), handle.getTableName());
+            // partition recovery
+            findAndRecoverPartitions(handle.getConnectorSession(),
+                    targetLocation,
+                    filesWritten.keySet(),
+                    partitionsKnown,
+                    table);
+        }
+    }
+
+    private void findAndRecoverPartitions(ConnectorSession session,
+                                          Path tableLocation,
+                                          Set<String> partitionsWritten,
+                                          Set<String> partitionsKnown,
+                                          final Table table) throws TException
+    {
+        List<Partition> commitBatch = new ArrayList<Partition>();
+        int recovered = 0;
+
+        for (String partition : partitionsWritten) {
+            if (!partitionsKnown.contains(partition)) {
+                // create the partition location using tableLocation and the values
+                Path partLocation = new Path(tableLocation, partition);
+                Partition tpart = getMetastore().createPartition(table.getDbName(), table.getTableName(), getPartitionValues(partition), null, table, partLocation.toString());
+                commitBatch.add(tpart);
+                recovered++;
+
+                if (commitBatch.size() >= partionCommitBatchSize) {
+                    getMetastore().addPartitions(commitBatch, table.getDbName(), table.getTableName());
+                    commitBatch.clear();
+                }
+            }
+        }
+
+        if (commitBatch.size() > 0) {
+            getMetastore().addPartitions(commitBatch, table.getDbName(), table.getTableName());
+            commitBatch.clear();
+        }
+
+        log.info("Recovered " + recovered + " partitions");
+    }
+
+    private void moveInsertIntoData(HiveInsertTableHandle handle, Map<String, List<String>> filesWritten) throws IOException
+    {
+        ExecutorService executor = Executors.newFixedThreadPool(
```

Comment:
Should this threadpool be centrally managed (guiced in)? If there are multiple insert statements, they will all create their own threadpools?